### ISSUE TITLE
[FeatReq] Cancelable Denizen actions

### ISSUE THREAD

**Issue by spaceemotion** [Dandielo/DtlCitizensTrader#41]
It would be nice to have some kind of Denizen action called onTraderOpen and onTraderClose, that can be cancelled.
My idea: A trader that only opens his stock when you made a quest or set a denizen-flag. I would then be able to put some CHAT messages before (like "hey you finally got XYZ, feel free to buy some stuff" or "hey, i told you i won't trade with strangers, didn't i?").

Would be nice to have :)

**Comment by Dandielo** [comment]
Sure :) will try to add it :)  but might be not easy :P 

**EDIT:** This might take some time because waiting for an denizen API that could do this :D because there so far no Callbacks after a Action has ended :P

**Comment by Dandielo** [comment]
Started working on it, but as said it may take some time :P

**Comment by spaceemotion** [comment]
I am a denizen developer - if you tell me what you need i can implemented it :D

**Comment by Dandielo** [comment]
hmm... because there are some ways to achieve this, one of them is to "listen" a "trader click" trigger and then use dtltrader open/close command, 

another way is to create a <trader.clallback.open/close> tag, that would read some info from a specified flag. and if it will be true a scheduled statement would open or close the inventory, 

or (I dunno how generally actions work atm), allow to add on the action caller sth like a callback function, so if the actions ends, this function will be called (but this would be a denizen side work), 

btw your sitting on Denizen IRC? Because I don't know your nickname xd

**Comment by Dandielo** [comment]
Ok after a day full of laziness and a bit work :D 
You can use the "TRADER" command to open a traders inventory, atm only one argument "OPEN" that is not required and default :P

Also added TraderClickTrigger, better said "TraderTrigger" you can use it as a normal Trigger
or as an Action "On Trader"

added these tags:
     \* < trader.stock.sell.trigger>
     \* < trader.stock.sell.hint>
     \* < trader.stock.buy.trigger>
     \* < trader.stock.buy.hint>
     \* < trader.pattern>
     \* < trader.owner>
     \* < trader.wallet>
     \* < trader.wallet.balance>
     \* < transaction.trigger>
     \* < transaction.item>
     \* < transaction.item.price>
     \* < transaction.item.amount>
     \* < transaction.item.instock>
     \* < transaction.failure>

< trader.stock.<...>.trigger> are automatically replaced with a regex for all trader items in the specified stock
< trader.stock.<...>.hint> creates string that has each item (it's regex name) and it's price in the specified stock

Buy stock is still not done sorry :P 

< transaction.trigger> holds the item name that triggered the chat trigger, it's used by the transaction command, 
- TRANSACTION 'ITEM:<transaction.trigger>'

< trader...> tags are available always where, <transaction...> tags can be used only after the Transaction command, these tags provides additional transaction information.

Transaction command

`TRANSACTION ({SELL}|BUY) [ITEM:#(:#)] (QTY:#)`

QTY: should be used for different amounts :) 

Please report any bugs :)

**Comment by Dandielo** [comment]
waiting for a denizen issue fix, 

and for now these tags are change into 
- < trader.stock.sell-trigger>
  - < trader.stock.sell-hint>
  - < trader.stock.buy-trigger>
  - < trader.stock.buy-hint>

**Comment by spaceemotion** [comment]
yep thank you!
I am glad i hadn't started to work on the trader npc scripts yet ;)

**Comment by Dandielo** [comment]
This feature needs some debug info :D and feature requests so I know what you want :)

**Comment by Dandielo** [comment]
Wiki incoming in the next 2-3 days
